### PR TITLE
_describe API refactoring

### PR DESCRIPTION
**Pull request by u3ks**
Functional blockcount using groupby.

Potentially, it could be reimplemented with two calls to describe_reached:
 1. one with to aggregate the values on the neighbours graph
 2.one to aggregate the areas
 then return 1 / 2

Or it could be left to the user.

### CODE REVIEW

File: momepy/functional/_intensity.py [hunk 1]
```diff
@@ -111,3 +111,51 @@ def _calc_nodedensity(group, edges):
         summation_values = pd.Series(np.ones(nodes.shape[0]), index=nodes.index)
 
     return graph.apply(summation_values, _calc_nodedensity, edges=edges)
+
+
+def block_counts(
+    group_key: pd.Series | np.ndarray, graph: Graph, areas=None
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:117]:
```suggestion
    aggregation_key: pd.Series | np.ndarray, graph: Graph, areas=None
```

Comment (martinfleis) [reply momepy/functional/_intensity.py:117]: add typing to areas as well

File: momepy/functional/_intensity.py [hunk 2]
```diff
@@ -111,3 +111,51 @@ def _calc_nodedensity(group, edges):
...
+) -> pd.Series:
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:118]:
```suggestion
) -> Series:
```
consistency

**Review by martinfleis — CHANGES_REQUESTED** [review]

File: momepy/functional/_diversity.py [hunk 3]
```diff
@@ -226,6 +242,8 @@ def describe_reached(
     include_mode : False
         Compute mode along with other statistics. Default is False. Mode is
         computationally expensive and not useful for continous variables.
+    include_nunique : False
+        Compute number of uniue elements along with other statistics. Default is False.
```
Comment (martinfleis) [inline momepy/functional/_diversity.py:246]:
```suggestion
        Compute the number of unique elements along with other statistics. Default is False.
```

File: momepy/functional/_intensity.py [hunk 4]
```diff
@@ -111,3 +113,43 @@ def _calc_nodedensity(group, edges):
...
+
+    The unique number of aggregations within neighbours defined in ``graph``,
+    optionally divided by the area covered by the neighbours.
+
+    Adapted from :cite:`dibble2017`.
+
+    Parameters
+    ----------
+    aggregation_key: pd.Series
+        The group key that denotes group membership of ``graph`` elements.
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:130]: Can you say something like "for example..." to make this clearer?

File: momepy/functional/_intensity.py [hunk 5]
```diff
@@ -111,3 +113,43 @@ def _calc_nodedensity(group, edges):
...
+
+    Examples
+    --------
+    >>> tessellation_df['blocks_within_4'] = mm.block_counts(
+    ...                                       df_tessellation['bID'],
+    ...                                       graph_queen4,
+    ...                                       buildings_df['uID'])
+    """
+
+    results = describe(aggregation_key, graph, include_nunique=True)["nunique"]
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:149]: the overhead of computing the rest is negligible here? or shall we make the rest optional as well so we compute just `nunique`?

File: momepy/functional/_intensity.py [hunk 6]
```diff
@@ -111,3 +113,43 @@ def _calc_nodedensity(group, edges):
         summation_values = pd.Series(np.ones(nodes.shape[0]), index=nodes.index)
 
     return graph.apply(summation_values, _calc_nodedensity, edges=edges)
+
+
+def count_unique(aggregation_key: Series, graph: Graph, areas: Series = None) -> Series:
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:118]:
```suggestion
def count_unique(aggregation_key: Series, graph: Graph, area: Series = None) -> Series:
```

singular

File: momepy/functional/_intensity.py [hunk 7]
```diff
@@ -111,3 +113,43 @@ def _calc_nodedensity(group, edges):
...
+
+    Adapted from :cite:`dibble2017`.
+
+    Parameters
+    ----------
+    aggregation_key: pd.Series
+        The group key that denotes group membership of ``graph`` elements.
+    graph : libpysal.graph.Graph
+        A spatial weights matrix for the tessellations.
+    areas : Series, default None
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:133]:
```suggestion
    area : Series, default None
```

File: momepy/functional/_intensity.py [hunk 8]
```diff
@@ -111,3 +113,43 @@ def _calc_nodedensity(group, edges):
...
+    ...                                       df_tessellation['bID'],
+    ...                                       graph_queen4,
+    ...                                       buildings_df['uID'])
+    """
+
+    results = describe(aggregation_key, graph, include_nunique=True)["nunique"]
+
+    if areas is not None:
+        agg_areas = describe(areas, graph)["sum"]
+        results = results / agg_areas
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:153]:
```suggestion
    if area is not None:
        agg_areas = describe(area, graph)["sum"]
        results = results / agg_areas
```

File: momepy/functional/_intensity.py (hunk 5 above)
Comment (u3ks) [reply momepy/functional/_intensity.py:149]: i think so, running the compute_stats function takes around 1 sec for around 100k buildings

**Comment by martinfleis** [comment]
This will be superseded by` libpysal.Graph.describe`, correct?

**Comment by u3ks** [comment]
yes, i need to add the decorator and fix some of the failing tests though.

File: momepy/functional/_diversity.py [hunk 9]
```diff
@@ -45,181 +48,124 @@ def _interpolate(values, q):
...
-        np.min(values),
-        np.max(values),
-        np.sum(values),
-    ]
 
-    if include_mode:
-        results.append(_mode(values))
+def describe_agg(
+    y: np.ndarray | Series,
+    aggregation_key: np.ndarray | Series,
```
Comment (martinfleis) [inline momepy/functional/_diversity.py:69]: Can you use NDArray from the typing module for type hints?

File: momepy/functional/_diversity.py [hunk 10]
```diff
@@ -45,181 +48,124 @@ def _interpolate(values, q):
...
     ----------
-    grouper : pandas.GroupBy
-        Groupby Object which specifies the aggregations to be performed.
+    y : Series | numpy.array
+        A Series or numpy.array containing values to analyse.
+    graph_index : Series | numpy.array
+        The unique ID that specifies the aggregation
+        of ``y`` objects to ``graph`` groups.
+    result_index : pd.Index (default None)
+        An index that specifies how to order the results.
```
Comment (martinfleis) [inline momepy/functional/_diversity.py:95]: Can you explain in a sentence here why? I know that but user does not.

File: momepy/functional/_diversity.py [hunk 11]
```diff
@@ -45,181 +48,124 @@ def _interpolate(values, q):
...
+    result.index.names = result_index.names
 
-    return _compute_stats(grouper, q, include_mode)
+    return result
 
 
-def describe_reached(
+def describe_reached_agg(
     y: np.ndarray | Series,
     graph_index: np.ndarray | Series,
```
Comment (martinfleis) [inline momepy/functional/_diversity.py:156]: Same request on type hints as above

File: momepy/functional/_diversity.py [hunk 12]
```diff
@@ -45,181 +49,126 @@ def _interpolate(values, q):
...
+    # aggregate data
+    if q is None:
+        grouper = y.groupby(aggregation_key)
+    else:
+        grouper = _percentile_limited_group_grouper(y, aggregation_key, q=q)
+
+    stats = _compute_stats(grouper, to_compute=to_compute)
+
+    if result_index is None:
+        result_index = stats.index
```
Comment (martinfleis) [inline momepy/functional/_diversity.py:142]: Can we ensure this option is tested?

File: momepy/functional/_diversity.py [hunk 13]
```diff
@@ -307,26 +228,20 @@ def describe_reached(
...
 
-    else:
-        df_multi_index = y.to_frame().set_index(graph_index, append=True).swaplevel()
-        combined_index = graph.adjacency.index.join(df_multi_index.index).dropna()
+    if q is None:
         grouper = y.loc[combined_index.get_level_values(-1)].groupby(
             combined_index.get_level_values(0)
         )
+    else:
+        grouper = _percentile_filtration_grouper(y, combined_index, q=q)
```
Comment (martinfleis) [inline momepy/functional/_diversity.py:239]: This is also not tested at the moment.

**Review by martinfleis — APPROVED** [review]

**Comment by jGaboardi** [comment]
Anything blocking this merge?

**Comment by martinfleis** [comment]
@jGaboardi I was just waiting for CI to finish :)

**Pull request merged**